### PR TITLE
[BUG #2025] Ensure QR code data is properly parsed

### LINKED ISSUE #2025: [Wallet] QR code generation and parsing is incomplete
### User Story

As a user, I want to have my QR code recognized by other apps and be able to handle other QR code properly.

### Description

*Type*: Feature

Follow [EIP #67 ](https://github.com/ethereum/EIPs/issues/67)

### PR DESCRIPTION
**Pull request by jeluard**
fixes #2025

### Summary:
Properly parse and generate data embedded in QR code

### Steps to test:
- Open Status
- Navigate wallet / request
- QR code should be readable by others wallet (status included)
- Navigate wallet / send
- Recipient should be correctly read  from QR code generated by other wallet (status included)

status: ready

### CODE REVIEW

File: src/status_im/ui/screens/profile/qr_code/views.cljs [hunk 1]
```diff
@@ -34,10 +35,7 @@
                                                                             :height (.-height layout)}]))}
       (when (:width dimensions)
         [react/view {:style (styles/qr-code-container dimensions)}
-         [qr-code {:value (if amount?
-                            (prn-str {:address (get contact qr-source)
-                                      :amount  amount})
-                            (str "ethereum:" (get contact qr-source)))
+         [qr-code {:value (eip67/generate-uri (get contact qr-source) (if amount? {:value  amount}))
```
Comment (yenda) [inline src/status_im/ui/screens/profile/qr_code/views.cljs]: when

File: src/status_im/ui/screens/qr_scanner/events.cljs [hunk 2]
```diff
@@ -34,7 +35,7 @@
 (defn handle-qr-request
   [db [_ context data]]
   (when-let [handler (get-in db [:qr-codes context])]
-    (dispatch [handler context data])))
+    (dispatch [handler context (:address (eip67/parse-uri data))])))
```
Comment (yenda) [inline src/status_im/ui/screens/qr_scanner/events.cljs]: re-frame/dispatch

File: src/status_im/ui/screens/qr_scanner/views.cljs [hunk 3]
```diff
@@ -1,50 +1,42 @@
...
-               :hide-nav?        hide-nav?
-               :nav-action       (when modal
-                                   (act/back #(dispatch [:navigate-back])))}]]))
+    [react/view
+     [status-bar/status-bar]
+     [toolbar/toolbar {:title            title
+                       :background-color toolbar.styles/toolbar-background1
+                       :hide-nav?        hide-nav?
+                       :nav-action       (when modal
+                                           (act/back #(dispatch [:navigate-back])))}]]))
```
Comment (yenda) [inline src/status_im/ui/screens/qr_scanner/views.cljs]: action

File: src/status_im/ui/screens/qr_scanner/views.cljs [hunk 4]
```diff
@@ -1,50 +1,42 @@
...
-      [view st/rectangle
-       [image {:source {:uri :corner_left_top}
-               :style  st/corner-left-top}]
-       [image {:source {:uri :corner_right_top}
-               :style  st/corner-right-top}]
-       [image {:source {:uri :corner_right_bottom}
-               :style  st/corner-right-bottom}]
-       [image {:source {:uri :corner_left_bottom}
-               :style  st/corner-left-bottom}]]]]))
+     [camera/camera {:onBarCodeRead #(dispatch [:set-qr-code identifier (camera/get-qr-code-data %)])
```
Comment (yenda) [inline src/status_im/ui/screens/qr_scanner/views.cljs]: re-frame/dispatch

**Review by yenda — CHANGES_REQUESTED** [review]
Nice only minor details

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 5]
```diff
@@ -1,13 +1,13 @@
 (ns status-im.ui.screens.wallet.send.events
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/events.cljs:1]: fengshui? :)

**Comment by flexsurfer** [comment]
@jeluard but what about generating qr? to properly parse we need properly generate first, here https://github.com/status-im/status-react/blob/8d2992d1a41601596c42f46afcebbef7d3bbf6a8/src/status_im/ui/screens/wallet/request/views.cljs#L33

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 5 above)
Comment (jeluard) [reply src/status_im/ui/screens/wallet/send/events.cljs:1]: Weird

**Comment by jeluard** [comment]
@flexsurfer Right I missed this one!

**Comment by jeluard** [comment]
@yenda @flexsurfer Pushed changes according to your coments.

**Comment by yenda** [comment]
good point but one could argue that since it is the ethereum standard it will already work with other wallets QR code (Walleth comes to mind for instance) So this could be merged and the write part done in another PR

**Review by yenda — APPROVED** [review]

**Comment by flexsurfer** [comment]
@jeluard maybe it would be better to implement it completely because now amount always 0, qr should be regenerated when amount input is changing

**Review by flexsurfer — APPROVED** [review]

**Comment by asemiankevich** [comment]
Several issues found during testing :
- can't scan QR code using Android device (either with Wallet scanner or QR Code reader app). Assume it is known issue https://github.com/status-im/status-react/issues/2023 (at least partially)
- I receive `error Address is invalid` when I scan code generated on Android by wallet scanner opened on iOS
the code I generated in Android Wallet request flow
image

the error I receive when scanning code above by iOS device within Wallet Send flow
image

- [x] As improvement  to be implemented: adding the faulty address in the alert so user can understand what actually was scanned and why it is wrong

**Comment by rasom** [comment]
is this PR in progress? @jeluard please resolve conflicts

**Comment by asemiankevich** [comment]
Keep receiving error

image

**Comment by jeluard** [comment]
This QR code is generated using latest status:

capture du 2017-10-12 16-42-52

I can read this fine on Android.

**Pull request merged**